### PR TITLE
Do not localize dates that appear within code, hyperlinks and quote blocks

### LINKED ISSUE #25: Time is converted in preformatted text
Time is parsed and converted/displayed even if the text is in a preformatted or In-line code.

Values between these special characters should be left alone.

image

Running v0.1.1 on MM 5.21.0

### LINKED ISSUE #49: Time is converted in hyperlinks
Some hyperlinks may be misinterpreted as containing localizable dates, causing unexpected broken and/or partially-rendered hyperlinks.

For example:

`https://example.com/directory/test-2022-0900` renders as ```https://example.com/directory/test-`2022-0900` (5:22 AM GMT)``` for me locally.

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
Currently this plugin localizes dates within the full raw message text (that may include Markdown) as sent by a user.

Dates that appear within code blocks should not be localized, since they represent a form of quoted context that should be rendered as-is to recipients.  The same is true for blockquotes.

The same is true of hyperlinks, as reported in #49.

To detect and ignore those elements, this change uses the [Mattermost-flavoured fork of the `marked` Markdown processor](https://github.com/mattermost/marked) (as [used by `mattermost-webapp`](https://github.com/mattermost/mattermost-webapp/blob/88ca1d8c09254e68ef42ae7f0ed835455f447cc5/package.json#L47)) to filter them out from the text that is scanned for localizable dates.

#### Ticket Link
Resolves #25.
Resolves #49.

### CODE REVIEW

**Comment by jayaddison-collabora** [comment]
(note for reviewers: the unit test failure here should be resolved by #47)

**Comment by jayaddison-collabora** [comment]
Hrm - one more case that isn't yet handled correctly:

```
"```
test
one
two
three
1pm GMT tomorrow
four
```"
```

(ish - multi-backtick escape skills fail me)

Renders as:

```
test
one
two
three
`1pm GMT tomorrow` *(Thu, Jan 6 1:00 PM GMT)*
four
```

File: webapp/src/time.js [hunk 1]
```diff
@@ -12,8 +13,15 @@ chrono.casual.parsers = chrono.casual.parsers.filter((parser) => {
 });
 
 export function convertTimesToLocal(message, messageCreationTime, localTimezone, locale) {
+    const filterRenderer = new marked.Renderer();
+    filterRenderer.code = () => '';
+    filterRenderer.codespan = () => '';
+    filterRenderer.link = () => '';
```
Comment (jayaddison-collabora) [inline webapp/src/time.js:19]: It looks like there may be other cases that would make sense to ignore too; `blockquote` for example.  Perhaps it'd make sense to create an array of the renderer properties to override and assign to each of them in a loop.

**Comment by jayaddison-collabora** [comment]
@mkraft apologies for the repeat notifications; here's another, slightly more involved, merge request.

We deployed a version of `mattermost-plugin-walltime` at @collabora on 2020-01-05 and I'm yet to hear any reports of complaints / problems with time localization problems since then (that's not to say it hasn't happened, but I'm not yet aware of any problems there).

The code from that deployment is visible at https://gitlab.collabora.com/tools/mattermost/mattermost-plugin-walltime/-/commits/deploys/test-20220105 -- it includes most of the changes here albeit without the hyperlink and code block filtering.

**Comment by jayaddison-collabora** [comment]
Please note that the merge commit here (7b11ab03e82d24ed24a23198bd561e2f569c5015) involved some manual conflict resolution regarding the NPM `package-lock.json` file.

It wasn't possible to (easily) have the `npm` command-line tool resolve these automatically, because the version of the CLI tool I ran locally (v8.3.1) complained about an integrity hash change for the `marked` package dependency.

That's potentially related to some kind of change that NPM made regarding the calculation of integrity checksums.  There's a possible note about that [here](https://github.com/npm/cli/issues/2971#issuecomment-808435906) (I'm not really completely clear on the root cause of this, but have seen the same problem occur within other projects as well).

**Comment by jayaddison-collabora** [comment]
cc @mkraft @cpanato (apologies for the notification spam; this is the last of the pull requests I'll mention you on for now) - this is a change to prevent text within preformatted markdown blocks from being transformed by the `mattermost-plugin-walltime` plugin.

**Comment by jayaddison-collabora** [comment]
A personal update: I'm planning to resign from @collabora effective Friday 2022-04-01, so this pull request may go unmaintained from that point.

**Comment by jayaddison** [comment]
For anyone attempting to recover/scavenge from this pull request to resolve either of the linked issues: please consider updating the version pin from the current value of https://github.com/mattermost/marked/commit/6ca9a6b3f4bdd35dbf58d06f5e53369791e05915 to a more recent version -- such as https://github.com/mattermost/marked/commit/2ef7f28cc7718e3f551c4ce9ea75fdd7580c2008 that is the latest at the time-of-writing this.

**Pull request closed without merging**